### PR TITLE
Fix/utxo chaining

### PR DESCRIPTION
**Pull request by jcnelson**
This fixes miner UTXO chaining (#2358), and adds unit tests and integration tests to verify that missed block commits are now included in a miner's UTXO chain when determining its sortition weight.

Nearly all of this new code is tests and test infrastructure.  The change itself is less than 20 lines long.

### CODE REVIEW

File: testnet/stacks-node/src/neon_node.rs [hunk 1]
```diff
@@ -2120,13 +2160,52 @@ impl StacksNode {
...
 
-        let res = bitcoin_controller.submit_operation(op, &mut op_signer, attempt);
-        if !res {
-            if !config.node.mock_mining {
-                warn!("Failed to submit Bitcoin transaction");
-                return None;
-            } else {
-                debug!("Mock-mining enabled; not sending Bitcoin transaction");
+        let mut send_tx = true;
+        if cfg!(test) {
```
Comment (gregorycoppola) [inline testnet/stacks-node/src/neon_node.rs:2164]: I think this function is already doing a lot for the reader to follow, and they will now have to keep track of a new long cryptic `cfg!(test)` block.

any way to inject from the outside without putting it into the prod code using some kind of "for test" version of an interface?

an alternative would be to make a function in this file called `maybe_test_late_block_commit`.

at the end of the day if that's not worth it and you want to check this in it's fine with me.

File: src/chainstate/burn/operations/leader_block_commit.rs [hunk 2]
```diff
@@ -679,17 +679,43 @@ impl LeaderBlockCommitOp {
...
             };
-            if miss_distance > self.block_height {
-                return Err(op_error::BlockCommitBadModulus);
-            }
-            let intended_sortition = tx
-                .get_ancestor_block_hash(self.block_height - miss_distance, &tx_tip)?
-                .ok_or_else(|| op_error::BlockCommitNoParent)?;
+            let intended_sortition = match epoch.epoch_id {
+                StacksEpochId::Epoch21 => {
+                    // correct behavior
```
Comment (gregorycoppola) [inline src/chainstate/burn/operations/leader_block_commit.rs:693]: maybe "correct behavior, fixes issues/2358"

**Review by gregorycoppola — COMMENTED** [review]
one idea to consider but otherwise LGTM if the calculations are really correct :)

File: src/chainstate/burn/operations/leader_block_commit.rs [hunk 3]
```diff
@@ -679,17 +679,43 @@ impl LeaderBlockCommitOp {
...
+                    if self.block_height < burnchain.first_block_height {
+                        return Err(op_error::BlockCommitPredatesGenesis);
+                    }
+                    let sortition_height = self
+                        .block_height
+                        .saturating_sub(burnchain.first_block_height);
```
Comment (kantai) [inline src/chainstate/burn/operations/leader_block_commit.rs:699]:
```suggestion
                    let sortition_height = self
                        .block_height
                        .checked_sub(burnchain.first_block_height)
                        .ok_or_else(|| op_error::BlockCommitPredatesGenesis)?;
```

Using `checked_sub` is more idiomatic than a manual check + saturating sub.

File: testnet/stacks-node/src/neon_node.rs (hunk 1 above)
Comment (kantai) [reply testnet/stacks-node/src/neon_node.rs:2164]: I agree with this -- this function is already quite complex, and it seems like this is something that would be pretty easy to at split out to another function:

```rust
#[cfg(test)]
fn test_commit_fault_injection(op, op_signer, attempt, bitcoin_controller, block_height) -> bool {
...
}
#[cfg(not(test))]
fn test_commit_fault_injection(op, op_signer, attempt, bitcoin_controller, block_height) -> bool {
   true
}
...
let send_commit = test_commit_fault_injection(op, op_signer, attempt, bitcoin_controller, block_height);
if send_commit {
...
}
...
```

**Review by kantai — APPROVED** [review]
This LGTM, just some superficial comments.

File: testnet/stacks-node/src/neon_node.rs [hunk 4]
```diff
@@ -70,8 +71,112 @@ use super::{BurnchainController, BurnchainTip, Config, EventDispatcher, Keychain
...
+#[cfg(not(test))]
+fn fault_injection_delay_transactions(
+    _bitcoin_controller: &mut BitcoinRegtestController,
+    _cur_burn_chain_height: u64,
+    _stacks_block_burn_height: u64,
+    _op: &BlockstackOperationType,
+    _op_signer: &mut BurnchainOpSigner,
+    _attempt: u64,
+) -> bool {
+    false
```
Comment (kantai) [inline testnet/stacks-node/src/neon_node.rs:177]: Shouldn't this be `true`?

File: testnet/stacks-node/src/operations.rs [hunk 5]
```diff
@@ -36,12 +36,16 @@ impl BurnchainOpSigner {
 
     pub fn sign_message(&mut self, hash: &[u8]) -> Option<MessageSignature> {
         if self.is_disposed {
+            debug!("Signer is dsposed");
```
Comment (kantai) [inline testnet/stacks-node/src/operations.rs:39]: Typo

File: testnet/stacks-node/src/neon_node.rs (hunk 4 above)
Comment (jcnelson) [reply testnet/stacks-node/src/neon_node.rs]: Yup, fixed

File: testnet/stacks-node/src/operations.rs (hunk 5 above)
Comment (jcnelson) [reply testnet/stacks-node/src/operations.rs]: Fixed

**Review by kantai — APPROVED** [review]

**Review by gregorycoppola — APPROVED** [review]

**Pull request merged**